Dynamic Dashboard: Various small m2 fixes and improvements

hafizrahman opened the pull request:
## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds three main things:

### 1. Disable edit button when reloading all data during pull to refresh.
I find that when opening the Customize screen to early while all data is still being reloaded, then the state of shown cards in the Customize might be changing on its own (due to the fact that the underlying data is still being computed).

At the same time, when adding some edits in the screen (e.g: enabling or disabling a card), the edit will get removed automatically once all data is reloaded.

As a quick fix, this PR disables the Edit button when still reloading all data, in commit 137a84bc76635a54969f31d6ef39a8a598cdbdbb

(note: the issue above also happens on `viewWillAppear()` when the app first opens, butI haven't found a proper solution for that)

---

(The same fix for below was already added in an earlier PR https://github.com/woocommerce/woocommerce-ios/pull/12901)

~### 2. Duplicate inbox added in Customize.~

~This is a discovered issue that can be replicated by: ~
~1. Log out of the app and in again~
~2. Tap Add new section card~
~3. Notice that the Inbox option is not on the Customize screen.~
~4. Close the screen and pull to refresh the dashboard~
~5. Tap Edit, confirm that Inbox is duplicated on the Customize screen~

~The commit a5036e2978f5c16f2b2cf27c324260db380b213f added a fix so this doesn't happen anymore. Props to @itsmeichigo for working on the fix.~

---

(Decided not to include below:)

~### 3. Reduce amount of time saved cards are loaded.~
~Previously, saved card loading is done whenever `updateDashboardCards()` is called, which was too frequent since that function is called by `refreshDashboardCards()` that has a combine in it.~

~With this PR, the saved card is loaded once during init, then saved in `savedCards` variable. This variable is then updated during user saving action in Customize (via `didCustomizeDashboardCards()`) and during manual saving related to new notice logic, thus making it always be updated properly. The variable is then used for comparing purposes, instead of loading fresh all the time.  This is added in b0d9b2c091c171ef82b89c382928b9b0b2e6e84c~



## Testing instructions
### 1. Disable edit button when reloading all data during pull to refresh.
Simply do a pull-to-refresh and ensure Edit button is disabled until loading is done and it is enabled again.

hafizrahman committed:
Disabled Edit button if dashboard cards are currently reloading.

hafizrahman committed:
Merge branch 'trunk'

hafizrahman committed:
Refactor loading saved card steps to be more efficient.

Previously saved card loading is done whenever updateDashboardCards is called, which can be too frequent since that function s called by refreshDashboardCards that has a combine in it.
In this commit, the saved card is loaded once during init, then saved in `savedCards` variable. This variable is updated during user saving (didCustomizeDashboardCards), thus making it always be updated properly.

hafizrahman committed:
Fix inbox card being duplicated issue. props @itsmeichigo

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Properly set variables as private set

hafizrahman commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:
```diff
@@ -516,23 +525,21 @@ private extension DashboardViewModel {
         /// If no saved cards are found, display the default cards.
         if savedCards.isEmpty {
-            dashboardCards = updatedCards
+            dashboardCards = initialCards
         } else {
+            // Based on saved cards, preserve existing enabled state for all available cards.
+            // This is needed because even if a user already disabled an available card and saved it, in `initialCards`
+            // the same card might be set to be enabled. To respect user's setting, we need to check the saved state and re-apply it.
+            let updatedCards = initialCards.map { initialCard in
+                if let savedCard = savedCards.first(where: { $0.type == initialCard.type }),
+                   savedCard.availability == .show && initialCard.availability == .show {
+                    return initialCard.copy(enabled: savedCard.enabled)
+                } else {
+                    return initialCard
+                }
+            }
```
This might seem like a big change but it's essentially just updating the logic to no longer load saved cards anymore,  the rest is the same.

hafizrahman committed:
Update configure notice function.

hafizrahman committed:
More update for changes to configureNewCardsNotice()

hafizrahman committed:
Remove unneeded mainActor

hafizrahman committed:
Remove another unneeded mainActor

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:
```diff
@@ -149,13 +154,13 @@ final class DashboardViewModel: ObservableObject {
-    @MainActor
-    func handleCustomizationDismissal() async {
-        await configureNewCardsNotice()
+    func handleCustomizationDismissal() {
```
nit: Unit tests need to be updated to stop calling this method asynchronously.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:
```diff
@@ -48,6 +48,8 @@ final class DashboardViewModel: ObservableObject {
         dashboardCards.filter { $0.availability == .show && $0.enabled }
     }
 
+    private var savedCards: [DashboardCard] = []
```
I am concerned about adding a new property to save the cards. We also already have `var dashboardCards: [DashboardCard]`, and the new `savedCards` which will make the code harder to follow and debug. 

From the PR description, I think that we do this to "Reduce amount of time saved cards are loaded.". Was the previous logic causing any bugs? Would it be possible to explore another solution without introducing the `savedCards` property?

selanthiraiyan replied:
Suggestion: If the requirement is that the locally `savedCards` values to be ready in order to render the cards on screen, I think we should wait for the value to be available (instead of running a `Task` in `DashboardViewModel` init) and show a loading indicator if needed. This is because we can't be sure that the `Task` started in `init` will run/finish before we access the `savedCards` value. What do you think?

hafizrahman replied:
>  "Reduce amount of time saved cards are loaded.". Was the previous logic causing any bugs? Would it be possible to explore another solution without introducing the savedCards property?

So the issue I'm seeing when debugging is that this line:

```
let savedCards = await loadDashboardCards() ?? []
```

https://github.com/woocommerce/woocommerce-ios/pull/12899/files#diff-5e4b4001de415046921cf84e4ce2b6c16646ae1bb2067aa6d9ec1fbfa2a855e5L522

is called over and over even in just one pull-to-refresh action. I think it is because it's called by the combine in `refreshDashboardCard()`, so it's re-called each time a publisher is updated. However, the value of the saved cards do not change during pull-to-refresh, so it seems inefficient to me to load it multiple times only to get the same value each time. But it's not causing a bug per se.

hafizrahman replied:
Decided to revert the part about savedCards refactoring so this discussion can be resolved for now.

selanthiraiyan reviewed (commented):
Thanks for working on this, Hafiz! I left a couple of comments. Kindly let me know your thoughts.

hafizrahman committed:
Update unit test.

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift

hafizrahman committed:
Update unit test to include the new inbox eligibility checker.

hafizrahman committed:
Revert "Refactor loading saved card steps to be more efficient."

This reverts commit b0d9b2c091c171ef82b89c382928b9b0b2e6e84c.

hafizrahman committed:
Post-revert fixes.

hafizrahman committed:
Temporarily disable slow unit tests.

They work but also took 1s+ in CI. We're planning to refactor the logic in DashboardViewModel and that will affect these, so temporarily disabling first.

hafizrahman commented:
@selanthiraiyan Good for another review. So it doesn't become a blocker, I reverted the commit about refactoring `savedCards` logic, so this PR now is just for the point 1 & 2 in the PR description above. I have also disabled the slow unit tests as we discussed in Slack.

@itsmeichigo Also so this doesn't become a blocker, I don't add the "disabled Edit on viewDidLoad" [we discussed here](https://github.com/woocommerce/woocommerce-ios/pull/12925#issuecomment-2146934493). I'm working on it but will just make it a new PR instead.

hafizrahman committed:
Fix unit test.

selanthiraiyan reviewed (approved):
LGTM :shipit: 

---

I notice that you have updated the PR description to remove ` Reduce amount of time saved cards are loaded.`. It might also make sense to remove `Duplicate inbox added in Customize.` point as well.
